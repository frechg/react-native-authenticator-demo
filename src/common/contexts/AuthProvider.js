import React from 'react'
import * as SecureStore from 'expo-secure-store';
import { initialAuthState, authStateReducer } from '../authStateReducer';
import * as Api from '../../services/auth';

export const AuthContext = React.createContext();

export const AuthProvider = (props) => {
  const [authState, authDispatch] = React.useReducer(authStateReducer, initialAuthState);

  const getAuthState = async () => {
    try {
      const token = await SecureStore.getItemAsync('authToken');
      const username = await SecureStore.getItemAsync('username');
      authDispatch({type: 'RESTORE_TOKEN', token: token, username: username});
    } catch (error) {
      throw new Error(error);
    }
  };

  const signIn = async (data) => {
    await Api.signIn(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        throw new Error('Email or password are incorrect.');
      }
    })
  };

  const signUp = async (data) => {
    await Api.signUp(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        const jsonResponse = await response.json();
        throw new Error(jsonResponse.errors[0])
      }
    });
  };

  const signOut = async () => {
    try {
      await SecureStore.deleteItemAsync('authToken');
      await SecureStore.deleteItemAsync('username');
      authDispatch({type: 'SIGN_OUT'});
    } catch (error) {
      throw new Error(error);
    }
  };

  const contextValue = React.useMemo(() => {
    return {authState, getAuthState, signUp, signIn, signOut};
  }, [authState]);

  return (
    <AuthContext.Provider value={contextValue}>
      {props.children}
    </AuthContext.Provider>
  );
};