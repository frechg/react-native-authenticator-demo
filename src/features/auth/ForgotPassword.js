import React from 'react';
import {
    Alert,
    View,
    TouchableWithoutFeedback,
    Keyboard,
    KeyboardAvoidingView,
    Platform
} from 'react-native';
import { AuthContext } from '../../common/contexts/AuthProvider';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import { BasicTextInput } from '../../common/components/BasicTextInput';
import { BasicSubmitButton } from '../../common/components/BasicSubmitButton';
import { styles, formStyles } from '../../common/styles';

export const ForgotPassword = ({ navigation }) => {
  const { passwordReset } = React.useContext(AuthContext);

  const validation = Yup.object({
    email: Yup.string().email('Invalid email address').required('Required')
  });

  const handleSubmit = async (values, {setSubmitting}) => {
    await passwordReset(values)
    .then((response) => {
      if (response.ok) {
        Alert.alert(
          'Success',
          'Check your email for a link to reset your password.',
          [{
            text: 'Ok',
            onPress: () => navigation.navigate('Sign In')
          }]
        );
      } else {
        Alert.alert('Error', 'Password reset rquest failed.');
      }
    })
    .catch((error) => {
      Alert.alert('Error', error.message);
    });

    setSubmitting(false);
  }

  return (
    <KeyboardAvoidingView 
      behavior={Platform.OS === "ios" ? "padding" : "height"}
      style={styles.container}
    >
      <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
        <View style={formStyles.formWrapper}>
          <Formik
            initialValues={{email: ''}}
            validationSchema={validation}
            onSubmit={handleSubmit}
          >
            <View>
              <Field
                component={BasicTextInput}
                name='email'
                placeholder='Email'
                keyboardType='email-address'
              />
              <BasicSubmitButton title='Request password reset' />
            </View>
          </Formik>
        </View>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  );
}