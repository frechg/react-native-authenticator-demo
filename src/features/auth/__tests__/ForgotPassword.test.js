import * as React from 'react';
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react-native';

import { ForgotPassword } from '../ForgotPassword';
import { AuthContext } from '../../../common/contexts/AuthProvider';

describe('<ForgotPassword />', () => {
  test('Render ForgotPassword screen', () => {
    const passwordReset = jest.fn();
    render(
      <AuthContext.Provider value={{passwordReset}}>
        <ForgotPassword />
      </AuthContext.Provider>
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('User submits form with valid data', async () => {
    const passwordReset = jest.fn().mockResolvedValue(true);

    render(
      <AuthContext.Provider value={{passwordReset}}>
        <ForgotPassword />
      </AuthContext.Provider>
    );

    const email = 'person@example.com';
    const input = screen.getByPlaceholderText('Email');
    const button = screen.getByText('Request password reset');
    
    fireEvent.changeText(input, email);
    fireEvent.press(button);

    await waitFor(() => {
      expect(passwordReset).toHaveBeenCalledWith(JSON.stringify({email: email}));
      expect(screen.getByText('Check your email for a link to reset your password.')).toBeTruthy();
    });
  });

  test('User sees error when email is missing', async () => {
    const passwordReset = jest.fn();
    render(
      <AuthContext.Provider value={{passwordReset}}>
        <ForgotPassword />
      </AuthContext.Provider>
    );

    fireEvent.press(screen.getByText('Request password reset'));

    await waitFor(() => {
      expect(screen.getByText('Required')).toBeTruthy();
    });
  });

  test('User sees error when email is invalid', async () => {
    const passwordReset = jest.fn();
    render(
      <AuthContext.Provider value={{passwordReset}}>
        <ForgotPassword />
      </AuthContext.Provider>
    );

    fireEvent.changeText(screen.getByPlaceholderText('Email'), 'not_an_email');
    fireEvent.press(screen.getByText('Request password reset'));

    await waitFor(() => {
      expect(screen.getByText('Invalid email address')).toBeTruthy();
    });
  });

  test('User sees error when request fails', async () => {
    const passwordReset = jest.fn().mockResolvedValue(false);
    
    render(
      <AuthContext.Provider value={{passwordReset}}>
        <ForgotPassword />
      </AuthContext.Provider>
    );

    fireEvent.changeText(screen.getByPlaceholderText('Email'), 'person@example.com');
    fireEvent.press(screen.getByText('Request password reset'));

    await waitFor(() => {
      expect(screen.getByText('Sorry, something went wrong.')).toBeTruthy();
    });
  });
});