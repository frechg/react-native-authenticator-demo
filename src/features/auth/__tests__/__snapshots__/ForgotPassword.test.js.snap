// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPassword /> Render ForgotPassword screen 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "padding": 32,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "marginBottom": 24,
            "width": "100%",
          }
        }
      >
        <TextInput
          keyboardType="email-address"
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Email"
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "rgb(140,140,140)",
              "borderRadius": 0,
              "color": "rgb(60,60,60)",
              "fontSize": 18,
              "paddingBottom": 12,
              "paddingRight": 12,
              "paddingTop": 12,
              "width": "100%",
            }
          }
          value=""
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(40,40,40)",
            "borderRadius": 6,
            "opacity": 1,
            "padding": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(255,255,255)",
              "fontSize": 18,
            }
          }
        >
          Request password reset
        </Text>
      </View>
    </View>
  </View>
</View>
`;
